receiver/postgresqlreceiver/scraper.go

```diff
@@ -0,0 +1,294 @@
+			continue
+		}
+
+		defer dbClient.Close()
+
+		p.databaseSpecificMetricCollection(
+			now,
+			dbClient,
+			blocksRead,
+			databaseRows,
+			operations,
+		)
+	}
+
+	return md, nil
+}
+
+func (p *postgreSQLScraper) databaseSpecificMetricCollection(
+	now pdata.Timestamp,
+	client client,
+	blocksRead pdata.NumberDataPointSlice,
+	databaseRows pdata.NumberDataPointSlice,
+	operations pdata.NumberDataPointSlice,
+) {
+	// blocks read by table
+	blocksReadByTableMetrics, err := client.getBlocksReadByTable()
+	if err != nil {
+		p.logger.Error("Failed to fetch blocks read by table", zap.Error(err))
+		return
+	}
+
+	for _, table := range blocksReadByTableMetrics {
+		for k, v := range table.stats {
+			i, ok := p.parseInt(k, v)
+			if !ok {
+				continue
+			}
+
+			attributes := pdata.NewAttributeMap()
+			attributes.Insert(metadata.A.Database, pdata.NewAttributeValueString(table.database))
+			attributes.Insert(metadata.A.Table, pdata.NewAttributeValueString(table.table))
+			attributes.Insert(metadata.A.Source, pdata.NewAttributeValueString(k))
+			addToIntMetric(blocksRead, attributes, i, now)
+		}
+	}
+
+	// database rows & operations by table
+	databaseTableMetrics, err := client.getDatabaseTableMetrics()
+	if err != nil {
+		p.logger.Error("Failed to fetch database table metrics", zap.Error(err))
+		return
+	}
+	for _, table := range databaseTableMetrics {
+		for _, key := range []string{"live", "dead"} {
+			value := table.stats[key]
+			i, ok := p.parseInt(key, value)
+			if !ok {
+				continue
+			}
+
+			attributes := pdata.NewAttributeMap()
+			attributes.Insert(metadata.A.Database, pdata.NewAttributeValueString(table.database))
+			attributes.Insert(metadata.A.Table, pdata.NewAttributeValueString(table.table))
+			attributes.Insert(metadata.A.State, pdata.NewAttributeValueString(key))
+			addToIntMetric(databaseRows, attributes, i, now)
+		}
+
+		for _, key := range []string{"ins", "upd", "del", "hot_upd"} {
+			value := table.stats[key]
+			i, ok := p.parseInt(key, value)
+			if !ok {
+				continue
+			}
+
+			attributes := pdata.NewAttributeMap()
+			attributes.Insert(metadata.A.Database, pdata.NewAttributeValueString(table.database))
+			attributes.Insert(metadata.A.Table, pdata.NewAttributeValueString(table.table))
+			attributes.Insert(metadata.A.Operation, pdata.NewAttributeValueString(key))
+			addToIntMetric(operations, attributes, i, now)
+		}
+	}
+}
+
+func (p *postgreSQLScraper) databaseAgnosticMetricCollection(
+	now pdata.Timestamp,
+	client client,
+	databases []string,
+	commits pdata.NumberDataPointSlice,
+	rollbacks pdata.NumberDataPointSlice,
+	databaseSize pdata.NumberDataPointSlice,
+	backends pdata.NumberDataPointSlice,
+) {
+	// commits & rollbacks
+	xactMetrics, err := client.getCommitsAndRollbacks(databases)
+	if err != nil {
+		p.logger.Error("Failed to fetch commits and rollbacks", zap.Error(err))
+		return
```

Comment:
Can we continue on to collect database size and backends? If so, we should probably delegate each of the three major blocks in this function into a separate function so we can return early only for that scope of collection.

This is actually the use case for [PartialScrapeError](https://github.com/open-telemetry/opentelemetry-collector/blob/db4aa87dcf261c65478a9925b8fcc561ebf8e096/receiver/scrapererror/partialscrapeerror.go#L21), which is a pattern we'll need to adopt eventually. It'd be great if you could take a look at using it here. There are some examples of it in use in the [hostmetricsreceiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/hostmetricsreceiver/internal/scraper) scrapers.

Reply:
For now, I challenge you to use the PartialScrapeError pattern for the high level "failed to collect" errors, but eventually we will need to use it all the way down to individual "failed to parse" errors as well.